### ISSUE TITLE
Inconsistent masking of Secrets when action fails

### ISSUE THREAD

**Issue by erzz** [google-github-actions/deploy-cloudrun#328]
### TL;DR

In the case of a failing deployment - the printing of the deployment command exposes secrets by not masking them in job output.

I don't know if the fact that we are using the action as part of a reusable workflow is contributing but worth mentioning and included in my example below.

### Expected behavior

Any secrets should be masked in job output

### Observed behavior

Secrets are not masked

### Action YAML

**Reusable workflow:**
```yaml
name: Deploy Cloud Run

on:
  workflow_call:
    inputs:
      # <---------- SELECT AUTHENTICATION METHOD --------->
      gcp-sa-auth:
        required: false
        type: boolean
        description: "Set to true to enable GCP Service Account Key authentication"
        default: false
      gcp-oidc-auth:
        required: false
        type: boolean
        description: "Set to true to enable GCP OIDC authentication"
        default: false
      # <--------------- CLOUD RUN OPTIONS --------------->
      cr-service-name:
        required: false
        type: string
        description: "The name to give to the cloud run service"
        default: ${{ github.event.repository.name }}
      cr-image:
        required: true
        type: string
        description: "The full registry + container image path to deploy from"
      cr-region:
        required: true
        type: string
        description: "The GCP region in which to deploy the service"
        default: ""
      cr-suffix:
        required: false
        type: string
        description: "Optional suffix to apply to the revision name (GCP generates if not set)"
        default: ""
      cr-flags:
        required: false
        type: string
        description: "Additional cloud run flags to apply during deployment"
        default: "--port 8080 --cpu 1 --memory 1024Mi --timeout 5m --concurrency 80 --min-instances 0 --max-instances 1 --no-allow-unauthenticated"
      # <----------- GITHUB ENVIRONMENT OPTIONS ---------->
      gh-env-name:
        required: false
        type: string
        description: "The name to give to the environment created in Github"
        default: ${{ github.ref_name }}
      gh-env-url:
        required: false
        type: string
        description: "Override the environment URL. If unset sets to the cloud run service's url"
    secrets:
      # <------------------ OIDC AUTH -------------------->
      wip:
        required: false
        description: "The workfload identity provider to use for OIDC auth"
      service-account:
        required: false
        description: "The service account to impersonate when using OIDC auth"
      # <----------------- SA KEY AUTH ------------------->
      service-account-key:
        required: false
        description: "The service account key to use for service account authentication"
      # <------------------ CLOUD RUN -------------------->
      cr-project-id:
        required: true
        description: "The GCP project in which to deploy the Cloud Run service"
      cr-env-vars:
        required: false
        description: "Comma separated list of KEY=value environment variables to be applied to the service"
    outputs:
      url:
        description: "The URL of the deployed Cloud Run service"
        value: ${{ jobs.deploy.outputs.url }}
      ref-slug:
        description: "A URL sanitized version of the github ref"
        value: ${{ jobs.deploy.outputs.ref-slug }}
      short-sha:
        description: "Captures the short SHA for use in this or later workflow jobs"
        value: ${{ jobs.deploy.outputs.short-sha }}

jobs:
  deploy:
    name: Deploy to Cloud Run
    runs-on: ubuntu-latest
    outputs:
      url: ${{ steps.url.outputs.url }}
      short-sha: ${{ env.GITHUB_SHA_SHORT }}
      ref-slug: ${{ env.GITHUB_REF_SLUG_URL }}
    environment:
      name: ${{ inputs.gh-env-name }}
      url: ${{ steps.url.outputs.url }}
    steps:
      - name: Checkout the code
        uses: actions/checkout@v3

      - name: Slugify github variables
        uses: rlespinasse/github-slug-action@4.2.2

      - name: Authenticate to GCP (SA Key)
        if: ${{ inputs.gcp-sa-auth }}
        uses: google-github-actions/auth@v0
        with:
          credentials_json: ${{ secrets.service-account-key }}

      - name: Authenticate to GCP (OIDC)
        if: ${{ inputs.gcp-oidc-auth }}
        uses: google-github-actions/auth@v0
        with:
          workload_identity_provider: ${{ secrets.wip }}
          service_account: ${{ secrets.service-account }}

      - name: Cloud Run
        id: deploy
        uses: google-github-actions/deploy-cloudrun@main
        with:
          project_id: ${{ secrets.cr-project-id }}
          service: ${{ inputs.cr-service-name }}
          image: ${{ inputs.cr-image }}
          region: ${{ inputs.cr-region }}
          suffix: ${{ inputs.cr-suffix }}
          env_vars: ${{ secrets.cr-env-vars }}
          flags: ${{ inputs.cr-flags }}

      # Required to handle override of deployment URL
      - name: Set Deployment URL
        id: url
        run: |
          if [ "${{ inputs.gh-env-url }}" == "" ]; then
            echo "Setting url output to ${{ steps.deploy.outputs.url }}"
            echo "::set-output name=url::${{ steps.deploy.outputs.url }}"
          else
            echo "Setting url output to ${{ inputs.gh-env-url }}"
            echo "::set-output name=url::${{ inputs.gh-env-url }}"
          fi
```

**Consuming project's workflow.**

```yaml
deploy:
    uses: some-org/workflows/.github/workflows/deploy-cloudrun.yml@main
    needs: build
    with:
      gcp-sa-auth: true
      cr-service-name: ${{ github.ref_name }}-${{ github.event.repository.name }}
      cr-image: ${{ needs.build.outputs.image-name }}:${{ needs.build.outputs.image-tag }}
      cr-region: europe-north1
      cr-suffix: ${{ needs.build.outputs.short-sha }}
      cr-flags: '--port 8080 --cpu 1 --memory 2048Mi --timeout 10m --concurrency 80 --min-instances 1 --max-instances 1 --service-account some-service-account@some-gcp-project.iam.gserviceaccount.com --vpc-connector some-vpc-connector --no-cpu-throttling'
    secrets:
      service-account-key: ${{ secrets.SOME_SA_KEY }}
      cr-project-id: some-gcp-project
      cr-env-vars: API_USERNAME=${{ secrets.DEV_API_USERNAME }}, API_PASSWORD=${{ secrets.DEV_API_PASSWORD }}
```

### Log output

```text
Run google-github-actions/deploy-cloudrun@main
/usr/bin/tar xz --warning=no-unknown-keyword --overwrite -C /home/runner/work/_temp/2992a0a1-967e-4956-9a8c-097c7b24e7c4 -f /home/runner/work/_temp/ff8e5f8a-3635-47fc-8b9e-b9e98f5d5837
Running: gcloud run deploy rv8/355590ccfeae9030-some-project --image eu.gcr.io/***/some-project:rv8-355590ccfeae9030-411cbef --quiet --platform managed --region europe-north1 --update-env-vars API_USERNAME=VISIBLE_IN_PLAIN_TEXT,API_PASSWORD=VISIBLE_IN_PLAIN_TEXT --revision-suffix 411cbef --port 8080 --cpu 1 --memory 2048Mi --timeout 10m --concurrency 80 --min-instances 1 --max-instances 1 --service-account some-service-account@***.iam.gserviceaccount.com --vpc-connector some-vpc-connector --no-cpu-throttling --project *** --format json
Error: failed to execute gcloud command `gcloud run deploy rv8/355590ccfeae9030-some-project --image eu.gcr.io/***/some-project:rv8-355590ccfeae9030-411cbef --quiet --platform managed --region europe-north1 --update-env-vars API_USERNAME=VISIBLE_IN_PLAIN_TEXT,API_PASSWORD=VISIBLE_IN_PLAIN_TEXT --revision-suffix 411cbef --port 8080 --cpu 1 --memory 2048Mi --timeout 10m --concurrency 80 --min-instances 1 --max-instances 1 --service-account some-service-account@***.iam.gserviceaccount.com --vpc-connector some-vpc-connector --no-cpu-throttling --project *** --format json`: ERROR: (gcloud.run.deploy) Invalid resource name [rv8/355590ccfeae9030-some-project]. The name must use only lowercase alphanumeric characters and dashes, cannot begin or end with a dash, and cannot be longer than 63 characters.
```
### Additional information

As you can see - some secrets were correctly masked such as project_id input but specifically those passed via the env_vars input were not and are VISIBLE_IN_PLAIN_TEXT

**Comment by sethvargo** [comment]
Hi @erzz 

Thank you for opening an issue. GitHub Actions (the platform) is what actually performs secret scrubbing. Two things here:

1. I would recommend opening a security issue with GitHub. It looks like, from your example, it's possible to exfiltrate secrets from a shared workflow because the "secret" provenance is lost.
2. I would suggest switching to the native [Cloud Run Secret Manager integration](https://cloud.google.com/run/docs/configuring/secrets). That would prevent this type of leak because the secrets are resolved at runtime instead of deploy time.

I don't know if there's anything we can do in our action, since you're passing secrets via environment variables. It would not be safe for us to parse and mark values as secret, since not all environment variables are secret.

**Comment by erzz** [comment]
Hi @sethvargo - thanks for getting back!

1. I think you are perhaps correct - that a secret (reusable workflow) --> input (action) is not honoured in their scrubbing mechanism
2. Agreed! This was an older project started before secret manager was available in CR

Alright lets close this for now and I will raise an issue with Github. Thanks!
